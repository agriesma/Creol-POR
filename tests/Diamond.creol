/*
 * Diamond.creol
 *
 * This test case is used to analyze how Diamond-shaped inheritance is going
 * to be executed in Creol.
 */
class A(a: Int)
begin
  op init == skip
  op run == skip
  with Any
    op m(out r: Int) == r := a
end

class B(b: Int) inherits A(b)
begin
  op init == assert a = b
  op run == skip
  with Any
    op n(out r: Int) == r := b
end

class C(c: Int) inherits A(c)
begin
  op init == assert a = c
  op run == skip
  with Any
    op n(out r: Int) == r := c
end

class D inherits B(1), C(2)
begin
  var x: Int
  var y: Int
  op init == skip
  op run == m(; x); n(; y); assert x = 1 && y = 2
end
