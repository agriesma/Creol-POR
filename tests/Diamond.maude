load interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "A" : Cl | Inh: noInh, Par: "a", Att: "a" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "m" : Mtdname | Param: noAid, Latt: "r" |-> null, Code: "r" ::= "a" ; return ( "r" ) >, Ocnt: 0 >

< "B" : Cl | Inh: "A" < "b" > , Par: "b", Att: "b" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "n" : Mtdname | Param: noAid, Latt: "r" |-> null, Code: "r" ::= "b" ; return ( "r" ) >, Ocnt: 0 >

< "C" : Cl | Inh: "A" < "c" > , Par: "c", Att: "c" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "n" : Mtdname | Param: noAid, Latt: "r" |-> null, Code: "r" ::= "c" ; return ( "r" ) >, Ocnt: 0 >

< "D" : Cl | Inh: "B" < int(1) >  ## "C" < int(2) > , Par: noAid, Att: "x" |-> null , "y" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( "sync.0" ! "this" . "m" ( emp ) )  ; ( "sync.0" ? ( "x" ) )  ; ( "sync.1" ! "this" . "n" ( emp ) )  ; ( "sync.1" ? ( "y" ) )  ; skip ; return ( emp ) >, Ocnt: 0 >

.
endm
