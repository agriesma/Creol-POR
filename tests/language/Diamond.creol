/*
 * Diamond.creol -- Corner cases of the Creol semantics.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * This test case is used to analyze how Diamond-shaped inheritance is going
 * to be executed in Creol.
 */
class A(a: Int)
begin
  op init == skip
  op run == skip
  op m(out r: Int) == r := a
end

class B(b: Int) inherits A(b)
begin
  op init == assert a = b
  op run == skip
  op n(out r: Int) == m(; r)
end

class C(c: Int) inherits A(c)
begin
  op init == assert a = c
  op run == skip
  op o(out r: Int) == m(; r)
end

class D inherits B(1), C(2)
begin
  var x: Int
  var y: Int
  op init == skip
  op run == n(; x); o(; y); assert x = 1 && y = 2
end
