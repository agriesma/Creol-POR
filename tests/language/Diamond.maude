load creol-interpreter
mod PROGRAM is
protecting CREOL-SIMULATOR .
op init : -> Configuration [ctor] .
eq init =
< "A" : Cl | Inh: noInh, Par: "a", Att: "a" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "m" : Mtdname | Param: noVid, Latt: "r" |-> null , "_" |-> null, Code: "r" ::= "a" ; return ( "r" ) >, Ocnt: 0 >

< "B" : Cl | Inh: "A" < "b" > , Par: "b", Att: "b" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "n" : Mtdname | Param: noVid, Latt: "r" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "m" (emp ) )  ; ( "label:0" ? ( "r" ) )  ; return ( "r" ) >, Ocnt: 0 >

< "C" : Cl | Inh: "A" < "c" > , Par: "c", Att: "c" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "o" : Mtdname | Param: noVid, Latt: "r" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "m" (emp ) )  ; ( "label:0" ? ( "r" ) )  ; return ( "r" ) >, Ocnt: 0 >

< "D" : Cl | Inh: "B" < int(1) >  ,, "C" < int(2) > , Par: noVid, Att: "x" |-> null , "y" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "n" (emp ) )  ; ( "label:0" ? ( "x" ) )  ; ( "label:1" ! "this" . "o" (emp ) )  ; ( "label:1" ? ( "y" ) )  ; skip ; return ( emp ) >, Ocnt: 0 >

.
endm
