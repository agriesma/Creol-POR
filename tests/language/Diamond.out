==========================================
reduce in PROGRAM : init .
result Configuration: 
< "A" : Cl | Inh: noInh, Par: "a", Att: "a" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "m" : Mtdname | Param: noVid, Latt: "r" |-> null, Code: "r" ::= "a" ;
    return("r") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "B" : Cl | Inh: "A" < "b" >, Par: "b", Att: "b" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "n" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "r" |-> null, Code:
    "label:0" ! "this" . "m"(emp) ; "label:0" ?("r") ; return("r") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "C" : Cl | Inh: "A" < "c" >, Par: "c", Att: "c" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "o" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "r" |-> null, Code:
    "label:0" ! "this" . "m"(emp) ; "label:0" ?("r") ; return("r") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "D" : Cl | Inh: "B" < int(1) > ## "C" < int(2) >, Par: noVid, Att: "x" |->
    null, "y" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "label:0" ! "this" . "n"(emp) ; "label:0" ?("x") ; "label:1" !
    "this" . "o"(emp) ; "label:1" ?("y") ; skip ; return(emp) >,
  Ocnt: 0 >
==========================================
rewrite in PROGRAM : init
main("D", emp) .
result Configuration: 
< ob("D0") : Qu | Size: 10, Dealloc: noDealloc, Ev: noMsg >

< ob("main") : Qu | Size: 1, Dealloc: noDealloc, Ev: noMsg >

< "A" : Cl | Inh: noInh, Par: "a", Att: "a" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "m" : Mtdname | Param: noVid, Latt: "r" |-> null, Code: "r" ::= "a" ;
    return("r") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "B" : Cl | Inh: "A" < "b" >, Par: "b", Att: "b" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "n" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "r" |-> null, Code:
    "label:0" ! "this" . "m"(emp) ; "label:0" ?("r") ; return("r") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "C" : Cl | Inh: "A" < "c" >, Par: "c", Att: "c" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "o" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "r" |-> null, Code:
    "label:0" ! "this" . "m"(emp) ; "label:0" ?("r") ; return("r") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "D" : Cl | Inh: "B" < int(1) > ## "C" < int(2) >, Par: noVid, Att: "x" |->
    null, "y" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "label:0" ! "this" . "n"(emp) ; "label:0" ?("x") ; "label:1" !
    "this" . "o"(emp) ; "label:1" ?("y") ; skip ; return(emp) >,
  Ocnt: 1 >

< ob("D0") : "D" | Att: "a" |-> int(2), "b" |-> int(1), "c" |-> int(2), "this"
    |-> ob("D0"), "x" |-> int(2), "y" |-> int(2),
  Pr: idle,
  PrQ: noProc, Lcnt: 11 >

< ob("main") : "NoClass" | Att: noSubst,
  Pr: idle,
  PrQ: noProc, Lcnt: 0 >
Bye.
